bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/TomlProvider.java

```diff
@@ -259,18 +259,89 @@ public boolean hasConfigs() {
+            throw new ConfigException(CONFIG_TYPE_NOT_SUPPORTED, variableName, mapType.toString());
+        }
+        visitedNodes.add(tomlValue);
+        TomlTableNode tomlTableValue = (TomlTableNode) tomlValue;
+        MappingInitialValueEntry.KeyValueEntry[] keyValueEntries =
+                new MappingInitialValueEntry.KeyValueEntry[tomlTableValue.entries().size()];
+        int count = 0;
+        for (Map.Entry<String, TopLevelNode> field : tomlTableValue.entries().entrySet()) {
+            String fieldName = field.getKey();
+            Object value =
+                    retrieveTypedValue(field.getValue(), variableName + "." + fieldName, mapType.getConstrainedType());
+            keyValueEntries[count] =
+                    new MappingInitialValueEntry.KeyValueEntry(StringUtils.fromString(fieldName), value);
+            count++;
+
+        }
+        return ValueCreator.createMapValue(mapType, keyValueEntries);
+    }
+
+    private Object retrieveTypedValue(TomlNode tomlValue, String variableName, Type type) {
+        if (isPrimitiveType(type.getTag())) {
+            return retrievePrimitiveValue(tomlValue, variableName, type);
+        }
+        switch (type.getTag()) {
+            case TypeTags.ARRAY_TAG:
+                return retrieveArrayValue(tomlValue, variableName, (ArrayType) type);
+            case TypeTags.RECORD_TYPE_TAG:
+                return retrieveRecordValue(tomlValue, variableName, type);
+            case TypeTags.MAP_TAG:
+                return retrieveMapValue(tomlValue, variableName, (MapType) type);
+            case TypeTags.TABLE_TAG:
+                return retrieveTableValue(tomlValue, variableName, (TableType) type);
+            case TypeTags.INTERSECTION_TAG:
+                Type effectiveType = ((IntersectionType) type).getEffectiveType();
+                if (effectiveType.getTag() == TypeTags.RECORD_TYPE_TAG) {
+                    return retrieveRecordValue(tomlValue, variableName, type);
+                }
+                return retrieveTypedValue(tomlValue, variableName, effectiveType);
+            default:
+                invalidTomlLines.add(tomlValue.location().lineRange());
+                throw new ConfigException(CONFIG_TYPE_NOT_SUPPORTED, variableName, type.toString());
+        }
+    }
+
     @Override
     public Optional<BTable<BString, Object>> getAsTableAndMark(Module module, VariableKey key) {
-        Type effectiveType = ((BIntersectionType) key.type).getConstituentTypes().get(0);
+        TableType tableType = (TableType) ((BIntersectionType) key.type).getEffectiveType();
+        Type constraintType = tableType.getConstrainedType();
+        if (constraintType.getTag() == TypeTags.INTERSECTION_TAG) {
+            constraintType = ((IntersectionType) constraintType).getEffectiveType();
+        }
+        if (constraintType.getTag() == TypeTags.RECORD_TYPE_TAG) {
+            tableType = (TableType) ((BIntersectionType) key.type).getConstituentTypes().get(0);
```

Comment:
Do we have tests for map types defined in another module?

Reply:
Also defined in another project module?

Reply:
added in last commit.